include/sys/async_notify.h

```diff
@@ -0,0 +1,361 @@
+ * async_notify_fetch_result().
+ *
+ * See async_notify_init_spinwait().
+ */
+#define ASYNC_NOTIFY_METHOD_SPINWAIT 1
+
+/*
+ * Select notification through @ref k_poll signal
+ *
+ * See async_notify_init_signal().
+ */
+#define ASYNC_NOTIFY_METHOD_SIGNAL 2
+
+/*
+ * Select notification through a user-provided callback.
+ *
+ * See async_notify_init_callback().
+ */
+#define ASYNC_NOTIFY_METHOD_CALLBACK 3
+
+#define ASYNC_NOTIFY_METHOD_MASK 0x03
+#define ASYNC_NOTIFY_METHOD_POS 0
+
+/**
+ * @brief Identify the region of async_notify flags available for
+ * containing services.
+ *
+ * Bits of the flags field of the async_notify structure at and above
+ * this position may be used by extensions to the async_notify
+ * structure.
+ *
+ * These bits are intended for use by containing service
+ * implementations to record client-specific information.  The bits
+ * are cleared by async_notify_validate().  Use of these does not
+ * imply that the flags field becomes public API.
+ */
+#define ASYNC_NOTIFY_EXTENSION_POS 2
+
+/*
+ * Mask isolating the bits of async_notify::flags that are available
+ * for extension.
+ */
+#define ASYNC_NOTIFY_EXTENSION_MASK (~BIT_MASK(ASYNC_NOTIFY_EXTENSION_POS))
+
+/**
+ * @brief Generic signature used to notify of result completion by
+ * callback.
+ *
+ * Functions with this role may be invoked from any context including
+ * pre-kernel, ISR, or cooperative or pre-emptible threads.
+ * Compatible functions must be isr-ok and not sleep.
+ *
+ * Parameters that should generally be passed to such functions include:
+ *
+ * * a pointer to a specific client request structure, i.e. the one
+ *   that contains the async_notify structure.
+ * * the result of the operation, either as passed to
+ *   async_notify_finalize() or extracted afterwards using
+ *   async_notify_fetch_result().  Expected values are
+ *   service-specific, but the value shall be non-negative if the
+ *   operation succeeded, and negative if the operation failed.
+ */
+typedef void (*async_notify_generic_callback)();
+
+/**
+ * @brief State associated with notification for an asynchronous
+ * operation.
+ *
+ * Objects of this type are allocated by a client, which must use an
+ * initialization function (e.g. async_notify_init_signal()) to
+ * configure them.  Generally the structure is a member of a
+ * service-specific client structure, such as onoff_client.
+ *
+ * Control of the containing object transfers to the service provider
+ * when a pointer to the object is passed to a service function that
+ * is documented to take control of the object, such as
+ * onoff_service_request().  While the service provider controls the
+ * object the client must not change any object fields.  Control
+ * reverts to the client:
+ * * if the call to the service API returns an error;
+ * * when operation completion is posted.  This may occur before the
+ *   call to the service API returns.
+ *
+ * Operation completion is technically posted when the flags field is
```

Comment:
@nordic-krch for https://github.com/zephyrproject-rtos/zephyr/pull/23601#discussion_r395066208 this is where the documentation was added.  I'll fix the spelling error in the next pass.